pkg/providers/openstack/openstack_auth.go

```diff
@@ -0,0 +1,97 @@
+package openstack
+
+import (
+	"fmt"
+	"github.com/emc-advanced-dev/unik/pkg/config"
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack"
+	"os"
+)
+
+type OpenStackHandle struct {
+	AuthClient *gophercloud.ProviderClient
+	Region     string
+}
+
+func ObtainCredentials(conf config.Openstack) (*gophercloud.AuthOptions, error) {
+	// Obtain credentials passed in config file
+	auth := gophercloud.AuthOptions{
+		IdentityEndpoint: conf.OSAuthUrl,
+		UserID:           conf.OSUserId,
+		Username:         conf.OSUserName,
+		Password:         conf.OSPassword,
+		TenantID:         conf.OSTenantId,
+		TenantName:       conf.OSTenantName,
+		DomainID:         conf.OSDomainId,
+		DomainName:       conf.OSDomainName,
+	}
+
+	// Override with environment variables
+	if v := os.Getenv("OS_AUTH_URL"); v != "" {
+		auth.IdentityEndpoint = v
+	}
+	if v := os.Getenv("OS_USER_ID"); v != "" {
+		auth.UserID = v
+	}
+	if v := os.Getenv("OS_USERNAME"); v != "" {
+		auth.Username = v
+	}
+	if v := os.Getenv("OS_PASSWORD"); v != "" {
+		auth.Password = v
+	}
+	if v := os.Getenv("OS_TENANT_ID"); v != "" {
+		auth.TenantID = v
+	}
+	if v := os.Getenv("OS_TENANT_NAME"); v != "" {
+		auth.TenantName = v
+	}
+	if v := os.Getenv("OS_DOMAIN_ID"); v != "" {
+		auth.DomainID = v
+	}
+	if v := os.Getenv("OS_DOMAIN_NAME"); v != "" {
+		auth.DomainName = v
+	}
+
+	// Validate
+	if auth.IdentityEndpoint == "" {
+		return nil, fmt.Errorf("Argument OS_AUTH_URL needs to be set.")
+	}
+	if auth.UserID == "" && auth.Username == "" {
+		return nil, fmt.Errorf("Argument OS_USER_ID or OS_USERNAME needs to be set.")
+	}
+	if auth.Password == "" {
+		return nil, fmt.Errorf("Argument OS_PASSWORD needs to be set.")
+	}
+	if auth.TenantID == "" && auth.TenantName == "" {
+		return nil, fmt.Errorf("Argument OS_TENANT_ID or OS_TENANT_NAME needs to be set.")
+	}
+
+	return &auth, nil
+}
+
+func GetOpenStackHandle(conf config.Openstack) (*OpenStackHandle, error) {
```

Comment:
Yeah, I guess it's fine to pass by value since it can never be `nil` and it can never change unless you restart the daemon.